### ISSUE TITLE
Language ruby bug/Syntax highlighting bug

### ISSUE THREAD

**Issue by mjerem34** [pulsar-edit/pulsar#236]
### Thanks in advance for your bug report!

- [X] Have you reproduced issue in [safe mode](https://pulsar-edit.dev/docs/atom-archive/faq/sections/what-does-safe-mode-do.html)?
- [X] Have you used the [debugging guide](https://pulsar-edit.dev/docs/atom-archive/hacking-atom/#debugging) to try to resolve the issue?
- [X] Have you checked our [FAQs](https://pulsar-edit.dev/docs/launch-manual/sections/faq/) to make sure your question isn't answered there?
- [X] Does your issue already [exist](https://github.com/issues?q=is%3Aissue+user%3Apulsar-edit+is%3Aopen)?
- [X] Have you checked you are on the latest release of Pulsar?

### What happened?

Hello,

I installed Pulsar just now and I'm very happy to see that it works much better than my old version of Atom and that my favorite editor still has a nice community behind it!

I switched all my configuration from Atom to Pulsar, everything works fine, except this: 
![image](https://user-images.githubusercontent.com/7813700/207923048-8afa0a57-879e-4f95-94f0-4938b89de985.png)

It looks like the theme detects some parts of the code as text. I deduced this by analyzing the theme package, in the file `/home/jeremy/.pulsar/packages/monokai-seti/styles/syntax-variables.less` line 2 there is this: `@syntax-text-color: #F8F8F2;`. And if I change the color here then everything becomes the chosen color.

I tried to look in the `language-ruby` packages on the Atom and Pulsar repo, but I didn't find any difference...

### Which OS does this happen on?

🐧 Arch based (Manjaro, Garuda, etc.)

### OS details

Manjaro Sikaris 22.0.0

### Which CPU architecture are you running this on?

64-bit(x86_64)

### What steps are needed to reproduce this?

Environment : 
Pulsar V1.63.2022121503
Electron V12.2.3
Chrome V89.0.4389.128
Node V14.16.0
Theme used: Monokai Seti Syntax

To test, copy/paste theses lines of code : 
```
# frozen_string_literal: true

describe Api::V1::Module::Controller, type: :request do
  let(:user) { create(:user, first_name: 'A') }

  describe '#index' do
    let(:query) { get '/api/v1/controller' }
    let!(:xxx) { create(:xxxx, xxx: '123') }
    let!(:xxx) do
      create(:xxxx, :xxxx, xxx: xxx, xxx: xxx, xxx: false)
    end
```

**Comment by confused-Techie** [comment]
Thanks so much for reporting this issue. This likely has something to do with the fact that we have bumped TreeSitter which, I believe, all grammars rely on.

This is a community package so if a change is needed there we might be limited, but lets hope we can find a solution.

**Comment by mjerem34** [comment]
Thank you for your help. I can't disable tree-sitter grammars because it doesn't appear in the settings. What I did was to use the language-ruby package for grammars. Like this:
1. In the config.cson file, add : 
```
"*":
  core:
    customFileTypes:
      "source.ruby.rails": [
        "rb"
      ]
```
2. Install the `language-rspec` package

With this my rspec and source files are well colored.

**Comment by mauricioszabo** [comment]
@mjerem34 it should be on "Core" settings:
![image](https://user-images.githubusercontent.com/138037/210258272-117ad876-e097-48b7-8c64-3ea07fd884b4.png)

Please, let me know if, disabling both `language-rspec` and the options in settings-view, the highlight comes back (it may need a restart of the editor). If it does, we'll need to track the issue on tree-sitter and see if:

* [ ] Does newer tree-sitter for Ruby removed colorization of RSpec files?
* [ ] Does our implementation of tree-sitter on Pulsar is doing the wrong thing?
* [ ] Can we fix this on Pulsar easily, or should we wait until we migrate to a new integration of tree-sitter (needed because of newer Electron and because our implementation is outdated)

Thanks in advance :)

**Comment by mjerem34** [comment]
Hello @mauricioszabo !
I'm sorry, I didn't see your message! 
I deactivated the tree-sitter in core as you said and it's a bit better, but not like before with Atom

**Comment by mauricioszabo** [comment]
So, an update: we're working on a newer implementation of tree-sitter. It's not usable yet, but we're advancing on this front:
![image](https://user-images.githubusercontent.com/138037/221201127-a9e5165b-c3a6-4853-868c-579882efa3cf.png)

Although it doesn't seem so, `Flame` and `Render` *are being* highlighted - they are just using a scope Pulsar still doesn't understand (old grammars used a scope called `class`, newer tree-sitter is using `constructor` for some reason - we need to fix this).

So, we're progressing on fixing this.

**Comment by savetheclocktower** [comment]
@AlexWayfer, if there are other aspects of Ruby syntax highlighting that you think are still lacking with the Modern Tree-sitter grammar, please do let us know on this ticket. Thanks!

**Comment by AlexWayfer** [comment]
> @AlexWayfer, if there are other aspects of Ruby syntax highlighting that you think are still lacking with the Modern Tree-sitter grammar, please do let us know on this ticket. Thanks!

Thanks for your attention and your work.

It's a lot better now.

I was confused with something, but was not sure what about.

Now I've detected this problem!

As you can see, methods in ruby should have blue color and symbols (`:foo`) kind of… cyan.

But in the current (v106) implementation called methods colored as cyan as well:

![image](https://github.com/pulsar-edit/pulsar/assets/6510020/2cb68fbc-e90e-45aa-aae6-a3ce0e2a89e2)

As you can see, only method definition and keyword like `new` have blue color.

But at the screenshot above you can see that any method, like `describe` or `subject` have blue color as well.

I hope it's understandable enough.

**Comment by AlexWayfer** [comment]
And a new report:

As you can see, `Telegram` is white here, but should not be:

![image](https://github.com/pulsar-edit/pulsar/assets/6510020/d1f30dc2-d397-4926-8879-9f8e19f5456b)

**Comment by savetheclocktower** [comment]
@AlexWayfer, I can confirm that #555 (specifically 4e52093cfcd050081710d1a0506528221194b2b2) solves the issue of method invocations being cyan when they ought to be blue.

I'll look into a fix for the namespace issue.

**Comment by savetheclocktower** [comment]
OK, that fix [was easy as well](https://github.com/pulsar-edit/pulsar/pull/555/commits/f3bf3ddfc5ef8c9a78316a2f52befd9dbf269ca0).

If you want to verify these fixes and don't mind working with a bleeding-edge Pulsar build, you can download the proper binary for Intel Linux [from this page](https://cirrus-ci.com/task/6050898347032576), though it might not be ready until CI is finished running on #555. If you don't want to bother with all that, that's also cool.

**Comment by AlexWayfer** [comment]
I've downloaded from [this page](https://cirrus-ci.com/task/4550494706728960) to check your recent commits, but I'm unable to launch it unfortunately:

```
~/Downloads/Programs/pulsar-1.105.2023070523
> ./pulsar
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
libva error: vaGetDriverNameByIndex() failed with unknown libva error, driver_name = (null)
[102532:0706/085431.344450:FATAL:gpu_data_manager_impl_private.cc(415)] GPU process isn't usable. Goodbye.
fish: Job 1, './pulsar ~/Projects/twitch_noti…' terminated by signal SIGTRAP (Trace or breakpoint trap)
```

The same for `Appimage` version.

I see nothing special in [the AUR package instructions](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=pulsar-bin), but I have not enough experience to… wait a minute.

OK, I've remembered my Atom experience and it works with `--disable-gpu-sandbox` flag. Weird.

And I've faced #531 at start, but it doesn't seem too important.

So, down to business: yes, I can confirm everything: blue methods, yellow namespaces and even `blocky` are working.

Thank you very much. I'm looking forward to the merge of this PR (beautiful number, btw).

**Comment by AlexWayfer** [comment]
@savetheclocktower a new catch, white class again:

![image](https://github.com/pulsar-edit/pulsar/assets/6510020/0031f196-af98-4c50-a2c3-b88a107ae55f)

The same in the root scope.

The code:

```ruby
RSpec::Matchers.define :be_boolean do
  match do |actual|
    expect(actual).to be_in [true, false]
  end
end
```

**Comment by savetheclocktower** [comment]
Quick fix. Added to #660. Thanks!

**Comment by bxkx** [comment]
I wonder if there's something off, still.
Here's a comparison between modern tree-sitter on, off and GitHub as reference

![comparison](https://github.com/pulsar-edit/pulsar/assets/15363535/036beb86-4cf0-4d2f-a145-9ebaa6d24bd4)

`something`  should be blue in all 3 occurences, shouldn't it? It seems like it get's thrown off by the missing parenthesis. But even then it looks like it's coloring too much yellow in general within a class?

![more2](https://github.com/pulsar-edit/pulsar/assets/15363535/2b9c4626-28d7-4a5c-a5bb-7daa754a8727)

I'm just a newb though so I'm sorry if I got something fundamentally wrong about this.

**Comment by savetheclocktower** [comment]
I think you've found a bug in the One Dark theme: `.meta.class` is styled as yellow (perhaps so it can try to target the name of the class when it's defined), but different scope names in the new grammar mean that things inside a class that would otherwise be plain text are yellow. That's a weird choice on that theme's part and I'll be sure to fix it. So try to imagine how this would look if all the yellow parts _inside_ the class body were actually white.

The reason why `something` isn't blue in all three occurrences is a judgement call of mine. In Ruby, all bare references to identifiers can possibly resolve to a method defined in scope — in which case it'd be calling that method — or to a variable declared earlier. There's no easy way to distinguish those two cases. My choices are (a) do what I've done here — make identifiers white unless they're unambiguously known to be method calls; or (b) highlight all identifiers as though they're method calls. I wouldn't say this is a settled issue; if people feel like like B is the better option, I could see us making a change here, or at last making it configurable.

Legacy Tree-sitter gets it wrong for another reason: it's targeting “`identifier` as direct child of `method`” so it can highlight the method name in the _definition_, but this also incorrectly matches the bare `something` on the next line.

**Comment by savetheclocktower** [comment]
@bxkx The theme bug will be fixed by #735. Thanks for reporting that!

**Comment by bxkx** [comment]
Looks like this seems to be an issue with other - non bundled - themes as well. At least it's easily fixable with those. But I just can't figure out how to edit the One Dark theme myself. I can't find the theme files on Arch and putting the change in styles.less doesn't do anything either.

Anyways, thanks for the explanation! I guess a toggle would be cool but I think both options are similar in the level of getting used to it-ness. I'd probably slightly prefer the way it currently is supposed to be.
